### ISSUE TITLE
[Question] Recommendation for writing messages to file?

### ISSUE THREAD

**Issue by humphrey09** [rcmdnk/shell-logger#4]
Dear Shell-Logger-Team, 

really like your shell-logger, nice work!

I would like to write the messages also to a logfile, can you give me a recommendation for that? In a first try, I wrote a simple wrapper for your methods, example for function "debug": 
```
SHELL_LOGGER_WRAPPER_LOGFILE_ABS_PATH="/tmp/shell_logger_$(date +"%Y-%m-%d_%I-%M-%p").log"

function logDebug () {
    debug "$*" | tee -a "${SHELL_LOGGER_WRAPPER_LOGFILE_ABS_PATH}"
}
```
But that way, the nice output coloring is being lost. 
Can you give me a hint/recommendation how to implement it properly?

Thanks in advance, 
humph

**Comment by rcmdnk** [comment]
This behavior can be chaned by `LOGGER_COLOR` option.

The default value is `auto`, which gives colors when output is given to stdout/stderr directly, and disables colors when output is passed to a pipe or a file.

If you set,

```
LOGGER_COLOR=true
```

then it gives colored output even if it is passed to a pipe or a file.

In this case, the content of the log file contains escape sequences for colors like:

```
^[[3m[2024/10/02 22:00:00][:28][DEBUG] abc^[[m                                  
```

(`^[[3m` and `^[[m`)

You can see colored log by such as `cat log`,

If you need to check the log file as a text file or edit the log,
it is better to keep the log plain (set `auto` or `never` to `LOGGER_COLOR`).

Otherwise, you can set `LOGGER_COLOR=true` and enjoy colored log always!

**Comment by humphrey09** [comment]
Dear @rcmdnk ,
thank you very much for your detailed reply!
You are right, the logfile itself should be kept plain/clean, without escape sequences. 

For writing a logfile, (in my copy) I (discarded the before mentioned wrapper as it interferes with the call stack and) added a short addition at the end of fn _logger (), which prints the logging information unformatted (tested for bash):

Section "Other global variables"
```
LOGGER_EXT_LOGFILE_ABS_PATH="/tmp/shell_logger_$(date +"%Y-%m-%dT%H-%M-%S").log"
LOGGER_EXT_LOGFILE_ENABLE="1"
```

End of function **_logger ()**:
```
  if [ "$LOGGER_EXT_LOGFILE_ENABLE" -eq "1" ];then
    printf "%s\n" "$msg" >> "${LOGGER_EXT_LOGFILE_ABS_PATH}"
  fi  
```
Might not be very beautiful, but works :)

Thanks again and best wishes, 
humph

**Comment by humphrey09** [comment]
One last addition: if also the error trace should be written to the log, the _echo_ 's in fn **error ()** could be written as follows:
```
# original
echo "Traceback (most recent call last):"
# conditional pipe:
echo "Traceback (most recent call last):" | { if [ "$LOGGER_EXT_LOGFILE_ENABLE" -eq "1" ]; then tee -a "${LOGGER_EXT_LOGFILE_ABS_PATH}"; fi; }
```
That would be one way of doing it (tested in bash).

**Comment by rcmdnk** [comment]
@humphrey09 

Thank you for the idea.

I implemented new options in v0.3.0:


| Variable Name            | Description                                                                                             | Default                                                   |
| :----------------------- | :------------------------------------------------------------------------------------------------------ | :-------------------------------------------------------- |
| LOGGER_FILE_OUTPUT       | If set, output is written to the file.                                                                  | "" (Not output file is defined)                           |
| LOGGER_FILE_ONLY         | If 1 and LOGGER_FILE_OUTPUT is set, not output will be given to stdout/stderr.                          | 0                                                         |
| LOGGER_FILE_LEVEL        | Output level for the file.                                                                              | Same as LOGGER_LEVEL                                      |
| LOGGER_FILE_APPEND       | If 1, output is appended to the file. Otherwise, the file is overwritten.                               | 0                                                         |

If you do something like:

```
#!/bin/bash

LOGGER_FILE_OUTPUT="log"

source /usr/local/etc/shell-logger

info Command succeeded.
err Command failed!
```

then, `Command succeeded.` and `Command failed` are shown in your terminal with colors, as stdout and stderr.

In addition, both outputs are written to "log" w/o colors.

**Comment by humphrey09** [comment]
Dear @rcmdnk ,

wow, thank you very much for implementing this so quickly! 
I will happily use v0.3.0 from now on 😀

Thanks again for your help!

Best wishes,
humph

**Comment by humphrey09** [comment]
One last note: there is an echo with file output in line 127, which creates a "hoge" file: I guess, this is from coding/debugging and could be removed.
https://github.com/rcmdnk/shell-logger/blob/master/etc/shell-logger#L127

Best wishes, 
humph

**Comment by rcmdnk** [comment]
oh...

It is fixed in v0.3.1.

Thanks!

**Comment by humphrey09** [comment]
Perfect, thank you very much for the fix!

And thank you for the shell-logger itself, I really like it!
